hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -390,15 +391,34 @@ def invariants(cls):
     def _repr_step(self, rule, data, result):
         output_assignment = ""
         if rule.targets:
+            number_of_results = (
+                len(result.values) if isinstance(result, MultipleResults) else 1
+            )
+            number_of_last_names = len(rule.targets) * number_of_results
+            last_names = self._last_names(number_of_last_names)
+            names_per_target = [
+                last_names[i :: len(rule.targets)] for i in range(len(rule.targets))
+            ]
             if isinstance(result, MultipleResults):
                 if len(result.values) == 1:
-                    output_assignment = f"({self._last_names(1)[0]},) = "
+                    output_names = [f"({name},)" for name in chain(*names_per_target)]
+                    output_assignment = " = ".join(output_names) + " = "
                 elif result.values:
-                    number_of_last_names = len(rule.targets) * len(result.values)
-                    output_names = self._last_names(number_of_last_names)
-                    output_assignment = ", ".join(output_names) + " = "
+                    # We prefer the per-result ordering - otherwise we get interleaved
+                    # ordering for duplicate targets when there are >1 results.
+                    # Note: "per-target" and "per-result" is only true in the abstract,
+                    # the actual value ordering may differ.
+                    names_per_result = list(zip(*names_per_target))  # transposed
+                    output_names = [", ".join(names) for names in names_per_result]
+                    if len(rule.targets) == 1:
+                        output_assignment = ", ".join(output_names) + " = "
+                    else:
+                        # Note special curly-bracket notation, it is meant to signal that this case
+                        # (only) does not have strict ordering of the result-to-target mapping
+                        output_assignment = "{" + "; ".join(output_names) + "} = "
```

Comment:
I'd prefer to keep the output as valid syntax, if we can.

Reply:
Sure! But I feel the "= ... = ... =" syntax used in other cases is misleading here, so perhaps just a comma-separated list (tuple) of all variables?

Reply:
I should really be sleeping at this point, but: if we take out the randomness, it seems like we have some cases for multiple-assignment (`a1 = b1 = state.whatever()`), some for tuple-unpacking-assignment, and some where there might be both.  In the "both" case, I think we really want to do that in multiple statements, so we have `a1, a2 = ...; b1, b2 = a1, a2`, rather than trying to cram it into a single line.

Reply:
Sounds like a plan :grin: For ease of later discussion, this is the situation at PR current state:

|  T↓\R→ | Single | Multiple-0 | Multiple-1 | Multiple-2
| -------- | ------- | -------- | ------- | ------- | 
|`a` | `a_0 = s.init()` | `s.init()` | `(a_0,) = s.init()` | `a_0, a_1 = s.init()` |
|`a, b, a` | `a_0 = b_0 = a_1 = s.init()` | `s.init()` | `(a_0,) = (b_0,) = (a_1,) = s.init()` | `{a_0, b_0, a_1, a_2, b_1, a_3, a_4, b_2, a_5} = s.init()` |

Feel free to adjust the table to the desired state. Apart from parenthesis-or-not quibbles I think it looks good _except_ the lower right cell.

Reply:
Ok, I looked into the code and it looks like there is no randomness in assigning values to bundles. I misinterpreted the random-draw side when a target is duplicated. Updated code to line-split the multi-multi case.

Reply:
Yes, that's what I intended, except that I think single-element tuples should be parenthesized.  It's just too easy to miss the trailing comma alone - see e.g. [ruff docs](https://docs.astral.sh/ruff/formatter/black/#single-element-tuples-are-always-parenthesized) or https://github.com/psf/black/issues/897.

(we'll also need to add a changelog entry, and wait for the current merge to finish)

Reply:
Yes, that's what I intended, except that I think single-element tuples should be parenthesized.  It's just too easy to miss the trailing comma alone - see e.g. [ruff docs](https://docs.astral.sh/ruff/formatter/black/#single-element-tuples-are-always-parenthesized) or https://github.com/psf/black/issues/897.

(we'll also need to add a changelog entry, and wait for the current merge to finish)

Reply:
Oops, I'd forgotten to git-add RELEASE.rst